implement aslBackwardsVisitor

katrinafyi opened the pull request:
This adds a visitor suitable for basic backwards analyses. In doing so, we refactor the visit\_\* functions to place them inside a class. This makes it possible to override one or more of the methods, as we do in aslBackwardsVisitor for visit\_stmts.

The backwards visitor has the same limitations as the existing forwards visitor. That is, merging results at control flow joins is difficult to do. However, for simple analyses, this can be worked around with careful use of enter/exit scope.

There is naming confusion in aslVisitor / aslForwardsVisitor / aslBackwardsVisitor which would be good to fix. aslVisitor defines the visit actions for specific AST nodes, whereas the aslForwardsVisitor/aslBackwardsVisitor applies these recursively in a particular order. Maybe these can be called aslForwardsTransform...?

katrinafyi committed:
working on backwards visitor

katrinafyi committed:
fix generic and restore compatibility.

katrinafyi committed:
add aslBackwardsVisitor

ncough commented:
Nice! My only other issue with the visitor is supporting transforms from individual statements into lists of statements. I'd be keen to get rid of the custom walks I keen implementing.

katrinafyi committed:
aslVisitor: change vstmt to return stmt list

BREAKING!
This change affects the signature of the Asl_visitor.visit_stmt method.
For compatibility, a visit_stmt_single method is provided with
equivalent behaviour to the old visit_stmt. There is also an added
helper function to convert visitActions on single statements to
visitActions on a list of statements.

Both of these compatibility helpers WILL THROW if used with a visitor
that returns non-singleton statement lists.

This gives the user the flexibility to insert new statements or delete
statements entirely. On the other hand, post-functions in
ChangeDoChildrenPost will need to handle lists of functions as well.

This follows the original CIL visitor: https://people.eecs.berkeley.edu/~necula/cil/api/Cil.cilVisitor.html

katrinafyi commented:
There is now a draft commit to support returning stmt list. However, it has not been tested at all beyond the existing transforms (which all operate on single statements).

Indeed, I think this will require a more careful implementation of the backwards visitor. One thought is that we need to make sure that we do not reverse statement lists before/after ChangeDoChildrenPost.

katrinafyi committed:
fix backwards visitor and rearrange code

it is no longer a good idea for the backwards and forwards visitors to have a subtyping relation.
